gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -1098,6 +1105,19 @@ private ContextAwareMeter getGroupMeterForDag(String dagId, String meterName, Ma
           group -> metricContext.contextAwareMeter(MetricRegistry.name(ServiceMetricNames.GOBBLIN_SERVICE_PREFIX, group, meterName)));
     }
 
+    /**
+     * Used to track metrics for different specExecutors to detect issues with the specExecutor itself
+     * @param dagNode
+     * @param meterName
+     * @param meterMap
+     * @return
+     */
+    private ContextAwareMeter getExecutorMeterForDag(DagNode<JobExecutionPlan> dagNode, String meterName, Map<String, ContextAwareMeter> meterMap) {
+      String executorName = dagNode.getValue().getSpecExecutor().getUri().toString();
+      return meterMap.computeIfAbsent(executorName,
+          executorUri -> metricContext.contextAwareMeter(MetricRegistry.name(ServiceMetricNames.GOBBLIN_SERVICE_PREFIX, executorUri, meterName)));
```

Comment:
to turn this on its head, the map you're carrying around and passing into here is basically just function memoization for this `computeIfAbsent` call.

up to you, but I might raise the level of abstraction by presenting this as a cache of `ContextAwareMeter`s that delegates read-through (initialization) to the `MetricRegistry`.

Reply:
Ah yeah that's a good point it is essentially a cache 😓 . I think we can leave it as is for now but when we refactor the metrics out of the DagManager this should have a different interface along with the other metrics for some consolidation/consistency.